CWE-476 (#946)
ossf/wg-best-practices-os-developers: docs/Secure-Coding-Guide-for-Python/CWE-703/CWE-476/README.md

## Non-Compliant Code Example - Verify that the object is not None before accessing or modifying its fields

`Noncompliant01.py` defines a function `is_valid_name()` that checks if a certain name is properly defined (contains exactly two words, both capitalized):

*[noncompliant01.py](noncompliant01.py):*

```py
# SPDX-FileCopyrightText: OpenSSF project contributors
# SPDX-License-Identifier: MIT
""" Non-compliant Code Example """

def is_valid_name(s: str) -> bool:
    """ Check if the input is a name with 2 capitalised Strings """
    names = s.split()
    if len(names) != 2:
        return False
    return s.istitle()

#####################
# Attempting to exploit above code example
#####################
name = is_valid_name(None) 

print(name)

```

The `is_valid_name()` function could get called with a null argument, which raises an `AttributeError` due to the `NoneType` object not having any `split` attribute.

## Compliant Code Example - Verify that the object is not None before accessing or modifying its fields

> The `compliant solution` includes the same implementation as the previous `non-compliant example` but now checks if the string is `None`.

Reviewer: I think some backticks are unnecessary here:
```suggestion
The `compliant solution` includes the same implementation as the previous non-compliant example but now checks if the string is `None`.
```
```suggestion
The compliant solution includes the same implementation as the previous `non-compliant example` but now checks if the string is `None`.
```

    """ Check if the input is a name with 2 capitalised Strings """
    if s is None:
        return False
    names = s.split()
    if len(names) != 2:
        return False
    return s.istitle()

#####################
# Attempting to exploit above code example
#####################
name = is_valid_name(None)

print(name)

```

## Non-Compliant Code Example - Confirm the object is an array or similar data structure before taking its length

`Noncompliant02.py` demonstrates the process of checking the number of students in a given classroom. In certain scenarios, such as a small school without a certain grade, the number of students in a classroom may legitimately be represented as `None`.

*[noncompliant02.py](noncompliant02.py):*

```py
# SPDX-FileCopyrightText: OpenSSF project contributors
# SPDX-License-Identifier: MIT
"""Non-compliant Code Example"""

def print_number_of_students(classroom: list[str]):
    """Print the number of students in a classroom"""
    print(f"The classroom has {len(classroom)} students.")

#####################
# Attempting to exploit above code example
#####################
print_number_of_students(None)

```

> The code example attempts to directly calling `len()` on a non-array value, but rather on `None`, which will raise a `TypeError`. Such unchecked dereferencing of a `None` value could result in an application crash or denial of service, impacting system availability.

Reviewer: This sentence sounds a bit odd. How about something like that?
```suggestion
The code example attempts to directly call `len()` on a non-array value, in this case on `None`, which will raise a `TypeError`. Such unchecked dereferencing of a `None` value could result in an application crash or denial of service, impacting system availability.
```
